### DISCUSSION TITLE
Seemingly arbitrary tag detection and non-detection

### DISCUSSION THREAD

**Discussion by derwok** [Zettlr/Zettlr#2341] [Ideas]
### Description

I used to have a tag `#d10` working in 1.8.9. With beta1 & beta2 I spotted this is not rendered as a tag anymore...
So I invested a little bit - it tunrs out that one-letter, two-letter and four-letter hashtags are possible. But not 3-leter tags.

<img width="174" alt="grafik" src="https://user-images.githubusercontent.com/4393698/130225585-ad99712e-2a99-4408-b3ac-73ceb2089553.png">

### Reproducing

Enter a tag with 3 letters, like e.g. `#doc` 

### Zettlr Version

Beta (if applicable)

### Specify version

2.0.0-beta.2 (2.0.0-beta.2)

### Your Platform

- [ ] Windows
- [X] macOS
- [ ] Linux

### Architecture

- [X] Intel 64bit
- [ ] ARM 64bit

### Operating System Version

MacOS Mojave 10.14.6 (18G9028)

### Additional Information

I love Zettlr!

**Comment by nathanlesage** [comment]
Actually, your example `#doc` would still be recognized as a valid tag. The cause of this is not a bug, but a feature: Since colors can be written in hexadecimal form as, e.g., `#1cb27e` (btw the Zettlr green), I manually disabled detection of anything that looks like a color rather than a tag. This way, if you want to just type a color (which at least I do often), it won't get recognized as a tag. So if you have a 3-, 6-, or 8-character string following a hashtag that solely consists of the letters a to f and the numbers 0 to 9, it will not detect this as a tag.

If I would disable that again, then people writing colors often would have a lot of non-tags listed. I could offer making that optional, but can't promise an ETA, since I fear that a lot of the code could become much slower just by making that optional …

**Comment by derwok** [comment]
Ah! That was the pattern - I could think of any usecase to resve it. And yes you're right with #doc. I was asuming by my shallow testing that all 3-letter tags don't work.

Now that I see your explanation I would nevertheless argue to keep the old behaviour... (not only because otherwise I have to restructure quite some of my tagging which conflicts with the new exception rule...) - but:

Here my 2cts...

1. First I think it will be rather complicated to explain to tag users: You might use tags to structure, but you have to avoid these and these tags... actually how many people are documenting colour in markdown outside of code blocks, anyhow?

2. I think the tagging feature is just too much on the "meta level" / power-feature side to bow in front of special content... We have many examples where content need to bow for markdown - and we have a well established means for that: escaping by **backslash** or code blocks.

So I would argue to go to he old behaviour of allowing all tags again without exception and tell the (probably small?) group of colour content users to either use \#123456 \#123 backslashing or fencing as code block.

My 2cts.

**Comment by nathanlesage** [comment]
I'm afraid backslashing won't work since hashtag signs are not special characters for Markdown per-se so the backslashes would also be retained in any exported output.

I have an idea, I'll just convert this thing into a discussion and have a discussion open up, and then we can implement whatever the discussion yields.

**Comment by georgd** [comment]
The idea of simply writing the colour codes looks very nice. However, IMO this can severely affect systematic tagging.

Couldn't this be solved by using two hash symbols for colour codes? So,  `##1cb27e` would be the notation for Zettlr green.

**Reply by nathanlesage** [reply]
I think this unnecessarily adds complexity and makes it even harder further down the line to work with these documents, I'm afraid :/

**Comment by nathanlesage** [comment]
What might be possible is to have users selectively tell Zettlr what should be detected as tags and what not. This would require some changes and we have to make sure that it's still fast. It basically involves converting the hard-coded regular expressions here into variable ones:

https://github.com/Zettlr/Zettlr/blob/develop/source/common/util/should-match-tag.ts

**Reply by georgd** [reply]
I think this could be a good compromise.

**Comment by derwok** [comment]
Hi, sorry to go a step back.. but to be honest: I did non fully get the use case why someone wants to put HTML/CSS color codes in markdown files? Is it to render colorful text? Or is it to write technical documentation where color codes are part of the to-be-read content? And: is it more needed while inside Zettlr? Or more needed when generating pandoc exports? Could you help me here? Or is there an existing issue/wish where this demand was generated (and I can learn the details)?

**Reply by nathanlesage** [reply]
It's actually just that I have many files in which I sometimes drop CSS codes to remember them. It's not really part of any Markdown specification, that's true, but that's why I explicitly mentioned that it's more something for me and I think this is why it's a good idea to implement this as a setting.

**Comment by kantrol** [comment]
As i throw a lot of python snippets in markdown files, I also get a big number of unwanted tags. Please add an option to ignore tag detection within code blocks and inline code. This also helps for CSS codes and avoids breaking markdown specification when exporting as the mentioned double hashtag or backslash option would.
